src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/FileIO/FileSystem.Portable.vb

```diff
@@ -0,0 +1,1522 @@
+                        Overwrite, UIOptionInternal.NoUI, UICancelOption.ThrowException)
+                Catch ex As Exception
+                    If (TypeOf ex Is IO.IOException OrElse TypeOf ex Is UnauthorizedAccessException OrElse
+                        TypeOf ex Is SecurityException OrElse TypeOf ex Is NotSupportedException) Then
+                        Exceptions.Add(SubFilePath, ex.Message)
+                    Else
+                        Throw
+                    End If
+                End Try
+            Next
+
+            ' Copy / move all the sub directories under this directory to target directory.
+            For Each SubDirectoryNode As DirectoryNode In SourceDirectoryNode.SubDirs
+                CopyOrMoveDirectoryNode(Operation, SubDirectoryNode, Overwrite, Exceptions)
+            Next
+
+            ' If this is a move, try to delete the current directory.
+            ' Using recursive:=False since we expect the content should be emptied by now.
+            If Operation = CopyOrMove.Move Then
+                Try
+                    IO.Directory.Delete(SourceDirectoryNode.Path, recursive:=False)
+                Catch ex As Exception
+                    If (TypeOf ex Is IO.IOException OrElse TypeOf ex Is UnauthorizedAccessException OrElse
+                        TypeOf ex Is SecurityException OrElse TypeOf ex Is IO.DirectoryNotFoundException) Then
+                        Exceptions.Add(SourceDirectoryNode.Path, ex.Message)
+                    Else
+                        Throw
+                    End If
+                End Try
+            End If
+        End Sub
+
+        Private Shared Sub CopyOrMoveFile(ByVal operation As CopyOrMove,
+                    ByVal sourceFileName As String, ByVal destinationFileName As String,
+                    ByVal overwrite As Boolean, ByVal showUI As UIOptionInternal, ByVal onUserCancel As UICancelOption
+                )
+            Debug.Assert(System.Enum.IsDefined(GetType(CopyOrMove), operation), "Invalid Operation!!!")
+
+            ' Verify enums.
+            VerifyUICancelOption("onUserCancel", onUserCancel)
+
+            ' Get the full path and remove any separator at the end. This will handle invalid path exceptions.
+            ' IMPORTANT: sourceFileName and destinationFileName should be used for throwing user exceptions ONLY.
+            Dim sourceFileFullPath As String = NormalizeFilePath(sourceFileName, "sourceFileName")
+            Dim destinationFileFullPath As String = NormalizeFilePath(destinationFileName, "destinationFileName")
+
+            ' Demand FileIOPermission on the given paths. (CONSIDER: Performance?)
+            ' We do it here instead of waiting for System.IO.File.Copy/Move to demand because
+            ' - Framework's code demand FileIOPermission firstly.
+            ' - We add the ability to move with overwrite flag, which leads to different behavior
```

Comment:
This reasoning should be no longer relevant -- they were working around some behavior in File.Copy and File.Move relating to CAS (code access security) which is all gone now. All this code should just call through to those methods, and do zero interop (which otherwise has to be re-implemented for Unix, yuck)